src/server/channel_store.cc

```diff
@@ -90,20 +129,111 @@ void ChannelStore::Fill(const SubscribeMap& src, const string& pattern, vector<S
 }
 
 std::vector<string> ChannelStore::ListChannels(const string_view pattern) const {
-  shared_lock lk{lock_};
   vector<string> res;
-  for (const auto& [channel, _] : channels_) {
-    if (pattern.empty() ||
-        stringmatchlen(pattern.data(), pattern.size(), channel.data(), channel.size(), 0) == 1) {
+  for (const auto& [channel, _] : *channels_) {
+    if (pattern.empty() || Matches(pattern, channel))
       res.push_back(channel);
-    }
   }
   return res;
 }
 
 size_t ChannelStore::PatternCount() const {
-  shared_lock lk{lock_};
-  return patterns_.size();
+  return patterns_->size();
+}
+
+ChannelStoreUpdater::ChannelStoreUpdater(ChannelStore* store, bool pattern, bool to_add,
+                                         ConnectionContext* cntx, uint32_t thread_id)
+    : store_{store}, pattern_{pattern}, to_add_{to_add}, cntx_{cntx}, thread_id_{thread_id} {
+}
+
+void ChannelStoreUpdater::Record(string_view key) {
+  ops_.emplace_back(key);
+}
+
+pair<ChannelStore::ChannelMap*, bool> ChannelStoreUpdater::GetTargetMap() {
+  auto* target = pattern_ ? store_->patterns_ : store_->channels_;
+
+  for (auto key : ops_) {
+    auto it = target->find(key);
+    DCHECK(it != target->end() || to_add_);
+    // We need to make a copy, if we are going to add or delete new map slot.
+    if ((to_add_ && it == target->end()) || (!to_add_ && it->second->size() == 1))
+      return {new ChannelStore::ChannelMap{*target}, true};
+  }
+
+  return {target, false};
+}
+
+void ChannelStoreUpdater::Modify(ChannelMap* target, string_view key) {
+  using SubscribeMap = ChannelStore::SubscribeMap;
+
+  auto it = target->find(key);
+
+  // New key, add new slot.
+  if (to_add_ && it == target->end()) {
+    target->emplace(key, new SubscribeMap{{cntx_, thread_id_}});
+    return;
+  }
+
+  // Last entry for key, remove slot.
+  if (!to_add_ && it->second->size() == 1) {
+    DCHECK(it->second->begin()->first == cntx_);
+    freelist_.push_back(it->second.Get());
+    target->erase(it);
+    return;
+  }
+
+  // RCU update existing SubscribeMap entry.
+  DCHECK(it->second->size() > 0);
+  auto* replacement = new SubscribeMap{*it->second};
+  if (to_add_)
+    replacement->emplace(cntx_, thread_id_);
+  else
+    replacement->erase(cntx_);
+
+  // The pointer can still be in use, so delay freeing it
+  // until the dispatch and update the slot atomically.
+  freelist_.push_back(it->second.Get());
+  it->second.Set(replacement);
+}
+
+void ChannelStoreUpdater::Apply() {
+  // Wait for other updates to finish, lock the control block and update store pointer.
+  auto& cb = ChannelStore::control_block;
+  cb.update_mu.lock();
+  store_ = cb.most_recent;
```

Comment:
`store_` variable looks redundant as a data member. 
you do not use it until calling Apply operation and then you just override it with most_recent. 
Seems like it could be local here and GetTargetMap can accept 'ChannelMap* src'  as an argument